### ISSUE TITLE
Problem with git-rev.sh and old Git versions (Ubuntu 14.04)

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#356]
I'm writing here as I'm analyzing.

I can run './default-configure', but when I run 'make' I get the following message:
```
make[1]: Entering directory `/home/julius/checkouts/dosemu2/src'
make[1]: *** No rule to make target `../--git-path', needed by `include/version.h'.  Stop.
make[1]: Leaving directory `/home/julius/checkouts/dosemu2/src'
make: *** [default] Error 2
```

Bash version is: 4.3.11(1)-release

I've bisected it to:
a28e6709d3525bf761388b511fc77dba23876af7 is the first bad commit
commit a28e6709d3525bf761388b511fc77dba23876af7
Author: Stas Sergeev <stsp@users.sourceforge.net>
Date:   Wed Mar 8 00:16:51 2017 +0300

    dash go to hell please [fixes #326]
    
    Please don't annoy me with the dash-related bug reports. :)

Changing /bin/bash back to /bin/sh allows me to build. I think the problem is not related to the script not working correctly with Bash though, but that something else can't handle its output on Ubuntu 14.04. With /bin/sh there's no proper output. With /bin/bash the output is:
```
--git-path 4fd24166c51b401f2df4a05d62bc048e4a972a43
```
I don't think the '--git-path' is supposed to appear in the output of the script. Running the commands manually:
```
julius@kruis ~/c/dosemu2> git rev-parse --symbolic-full-name HEAD
refs/heads/devel
julius@kruis ~/c/dosemu2> git rev-parse --git-path refs/heads/devel
--git-path
4fd24166c51b401f2df4a05d62bc048e4a972a4
```

I guess the '--git-path' in the second command is not expected. This seems to work for me:
```
julius@kruis ~/c/dosemu2> git rev-parse refs/heads/devel
4fd24166c51b401f2df4a05d62bc048e4a972a43
```
Git version is: 1.9.1.

Removing '--git-path' from line 4 and rerunning the compilation process:
```
make[1]: Entering directory `/home/julius/checkouts/dosemu2/src'
make[1]: *** No rule to make target `../4fd24166c51b401f2df4a05d62bc048e4a972a43', needed by `include/version.h'.  Stop.
make[1]: Leaving directory `/home/julius/checkouts/dosemu2/src'
make: *** [default] Error 2
```

So I didn't guess correctly what the intention of the script was.

**Comment by stsp** [comment]
This is a problem with git, we require git-2.0.
But I am not sure why you get the right things
with dash, this script has bash-specific constructs
that dash wont handle.
Of course this entire technique of using git refs
as the prerequisites in makefile, sucks a lot. But
I simply don't know how to properly trigger the
rebuild of some files after every commit.

**Comment by jschwartzenberg** [comment]
With /bin/sh there's no output on stdout (just stderr), so nothing gets appended which works.
What's the output of the script with git 2.0? Maybe I could find something to get the same output from 1.9?

Btw. the build is so fast in general that I do a git clean -xf before I rebuild after using git.

**Comment by stsp** [comment]
With recent enough git, the output looks like
this:
`/home/stas/src/dosemu2/.git/refs/heads/devel`

**Comment by bpranoto** [comment]
I also got the same problem in my ubuntu 12.04.

My solution is simply install latest git from PPA:

```
sudo apt-add-repository ppa:git-core/ppa
sudo apt-get update
sudo apt-get install git
```

**Comment by stsp** [comment]
So while I'd like to fix this, I don't know how.
If someone is familiar with git, then suggestions/patches
are welcome.

**Comment by bpranoto** [comment]
It is git problem, not dosemus...

**Comment by stsp** [comment]
> Do you really need the full path?

Unfortunately, for building in a secondary worktree.

**Comment by jschwartzenberg** [comment]
So what if you take the relative path from git and make it absolute through the shell-script?

**Comment by stsp** [comment]
Yes, this is exactly the main difficulty that my
script is trying to solve: find an absolute path
to the primary worktree even when building
inside a secondary one.
If you know the other ways of doing this, or,
more importantly, different solutions that will
not involve a lot of scripting, please let me know.

**Comment by jschwartzenberg** [comment]
It depends a bit on how the initial script is called. If you don't use the PATH to call it, then `dirname "$0"` should always return the directory where the script is.

**Comment by stsp** [comment]
I don't need the directory where the script is.
I need the directory with the primary git worktree.

**Comment by jschwartzenberg** [comment]
Then I don't understand your layout. I would expect the script to be in the same directory as the worktree as it's in the worktree.

**Comment by stsp** [comment]
The script in the current worktree finds the path
to the primary worktree.

**Comment by stsp** [comment]
OK, I suddenly recalled that old git doesn't
support worktrees anyway. So we don't need
to do anything special for it.
Should now be fixed.

**Comment by andrewbird** [comment]
Launchpad PPA build for 14.04 is now failing with this message
~~~
Enabling plugins: 
	sdl 
	fluidsynth 
	translate 
	translate/charsets 
	kbd_unicode 
	alsa 
	console 
	term 
	modemu 
	ladspa 
	libao 
	X 
	commands 
	gpm 
	midimisc 
	extra_charsets
make[1]: Leaving directory `/<<PKGBUILDDIR>>'
   dh_auto_build -a -O--parallel
fatal: Not a git repository (or any of the parent directories): .git
fatal: Not a git repository (or any of the parent directories): .git
make[1]: Entering directory `/<<PKGBUILDDIR>>'
fatal: Not a git repository (or any of the parent directories): .git
git version too old
fatal: Not a git repository (or any of the parent directories): .git
make[2]: Entering directory `/<<PKGBUILDDIR>>/src'
make[2]: *** No rule to make target `.git', needed by `include/version.h'.  Stop.
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory `/<<PKGBUILDDIR>>/src'
make[1]: *** [default] Error 2
make[1]: Leaving directory `/<<PKGBUILDDIR>>'
dh_auto_build: make -j4 returned exit code 2
make: *** [build-arch] Error 2
dpkg-buildpackage: error: debian/rules build-arch gave error exit status 2
--------------------------------------------------------------------------------
Build finished at 20170420-2107

Finished
--------

~~~

**Comment by stsp** [comment]
Should now be fixed.
Why are you using non-git builds though?

**Comment by andrewbird** [comment]
The PPA is importing the git repo into a bazaar repo for build. I'm not wedded to this, but at the time it was the only way to get it done. Maybe now that launchpad git support has matured a little, perhaps it might be possible to do a build from a native git repo.

**Comment by andrewbird** [comment]
Thanks, the PPA builds again.

**Comment by jschwartzenberg** [comment]
I can confirm this is solved here with Ubuntu 14.04.
